File: src/pratice/binarysearch/듣보잡1764.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 듣보잡1764 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, M;
    public static int cnt = 0;
    public static String[] A;
    public static String[] B;
    public static StringBuffer buffer = new StringBuffer();

    public static void main(String[] args) throws IOException {
        input();
        Arrays.sort(A);
        Arrays.sort(B);

        for (int i = 0; i < M; i++) {
            pro(B[i]);
        }
        System.out.println(cnt);
        System.out.println(buffer);
    }

    private static void pro(String name) {
        int L = 0;
        int R = N - 1;

        while(L <= R){
            int mid = (L + R) / 2;
            if(A[mid].equals(name)){
                cnt++;
                buffer.append(name).append("\n");
                return;
            }else if(A[mid].compareTo(name) < 0){
                L = mid + 1;
            }else{
                R = mid - 1;
            }
        }
    }

    private static void input() throws IOException {
        String[] split = br.readLine().split(" ");
        N = Integer.parseInt(split[0]);
        A = new String[N];
        M = Integer.parseInt(split[1]);
        B = new String[M];
        for (int i = 0; i < N; i++) {
            A[i] = br.readLine();
        }
        for (int i = 0; i <M;  i++) {
            B[i] = br.readLine();
        }
    }
}

File: src/pratice/binarysearch/듣보잡1764.kt
package pratice.binarysearch

fun main() {
    val (N, M) = readln().split(" ").map { it.toInt() }
    val a = mutableListOf<String>()
    val b = mutableListOf<String>()

    repeat(N) {
        a.add(readln())
    }
    repeat(M) {
        b.add(readln())
    }
    a.sort()
    b.sort()
    var cnt = 0
    val buffer = StringBuffer()
    for (name in b) {
        if (binarySearch(a, name)) {
            cnt++
            buffer.append(name).append("\n")
        }
    }
    println(cnt)
    println(buffer.toString())
}

private fun binarySearch(names: MutableList<String>, name: String): Boolean {
    var L = 0
    var R = names.size - 1
    while (L <= R) {
        val mid = (L + R) / 2
        if (names[mid] == name) {
            return true
        } else if (names[mid] < name) {
            L = mid + 1
        } else {
            R = mid - 1
        }
    }
    return false
}
